src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+        }
+
+        /// <summary>
+        /// Gets the method symbol the expression is enclosed within
+        /// </summary>
+        protected IMethodSymbol? GetMethodSymbolFromExpression(SemanticDocument annotatedSemanticDocument, TExpressionSyntax annotatedExpression, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = annotatedSemanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var methodDeclaration = annotatedExpression.FirstAncestorOrSelf<SyntaxNode>(node => IsMethodDeclaration(node), ascendOutOfTrivia: true);
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            return (IMethodSymbol)annotatedSemanticDocument.SemanticModel.GetDeclaredSymbol(methodDeclaration, cancellationToken)!;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        /// <returns>Dictionary tying all the invocations to each corresponding document</returns>
+        protected static async Task<ImmutableDictionary<Document, List<TInvocationExpressionSyntax>>> FindCallSitesAsync(SemanticDocument document,
+                                                                                                                         IMethodSymbol methodSymbol,
+                                                                                                                         CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct().ToImmutableArray()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            var list = new List<TInvocationExpressionSyntax>();
+            methodCallSites.Add(document.Document, list);
+            foreach (var refLocation in referencedLocations)
+            {
+                if (!methodCallSites.TryGetValue(refLocation.Document, out var invocations))
+                {
+                    invocations = new List<TInvocationExpressionSyntax>();
+                    methodCallSites.Add(refLocation.Document, invocations);
+                }
+                invocations.Add((TInvocationExpressionSyntax)refLocation.Location.FindNode(cancellationToken).Parent!);
```

Comment:
1. .GetRequiredParent for all these cases
2. do not blindly case to TInvocationExpressionSyntax.  the reference may not be an invocation. we def need tests of that.  for example just referencing a method with `this.foo` instead of `this.foo(...)`.